cirq-core/cirq/transformers/measurement_transformers.py

```diff
@@ -0,0 +1,191 @@
+                should create one new _MeasurementQid per qubit being measured
+                by that gate.
+        """
+        self._key = value.MeasurementKey.parse_serialized(key) if isinstance(key, str) else key
+        self._qid = qid
+
+    @property
+    def dimension(self) -> int:
+        return self._qid.dimension
+
+    def _comparison_key(self) -> Any:
+        return (str(self._key), self._qid._comparison_key())
+
+    def __str__(self) -> str:
+        return f"M('{self._key}', q={self._qid})"
+
+    def __repr__(self) -> str:
+        return f'_MeasurementQid({self._key!r}, {self._qid!r})'
+
+
+@transformer_api.transformer
+def defer_measurements(
+    circuit: 'cirq.AbstractCircuit', *, context: Optional['cirq.TransformerContext'] = None
+) -> 'cirq.Circuit':
+    """Implements the Deferred Measurement Principle.
+
+    Uses the Deferred Measurement Principle to move all measurements to the
+    end of the circuit. All non-terminal measurements are changed to
+    conditional quantum gates onto ancilla qubits, and classically controlled
+    operations are transformed to quantum controls from those ancilla qubits.
+    Finally, measurements of all ancilla qubits are appended to the end of the
+    circuit.
+
+    Optimizing deferred measurements is an area of active research, and future
+    iterations may contain optimizations that reduce the number of ancilla
+    qubits, so one should not depend on the exact shape of the output from this
+    function. Only the logical equivalence is guaranteed to remain unchanged.
+    Moment and subcircuit structure is not preserved.
+
+    Args:
+        circuit: The circuit to transform. It will not be modified.
+        context: `cirq.TransformerContext` storing common configurable options
+            for transformers.
+    Returns:
+        A circuit with equivalent logic, but all measurements at the end of the
+        circuit.
+    Raises:
+        ValueError: If sympy-based classical conditions are used, or if
+            conditions based on multi-qubit measurements exist. (The latter of
+            these is planned to be implemented soon).
+    """
+
+    context1 = context or transformer_api.TransformerContext()
+    circuit = transformer_primitives.unroll_circuit_op(circuit, deep=True, tags_to_check=None)
+    qubits_found: Set['cirq.Qid'] = set()
+    terminal_measurements: Set['cirq.MeasurementKey'] = set()
+    control_keys: Set['cirq.MeasurementKey'] = set()
+    for op in reversed(list(circuit.all_operations())):
+        gate = op.gate
+        if isinstance(gate, ops.MeasurementGate):
+            key = value.MeasurementKey.parse_serialized(gate.key)
+            if key not in control_keys and qubits_found.isdisjoint(op.qubits):
+                terminal_measurements.add(key)
+        for c in op.classical_controls:
+            control_keys.update(c.keys)
+        qubits_found.update(op.qubits)
```

Comment:
You can now use `transformers.synchronize_terminal_measurements.find_terminal_measurements` to find the terminal measurements in the given circuit. 

See https://github.com/quantumlib/Cirq/blob/master/cirq-core/cirq/transformers/synchronize_terminal_measurements.py#L25